ta/pkcs11/src/object.c

```diff
@@ -427,3 +427,552 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+
+	rc = serialargs_get_session_from_handle(&ctrlargs, client, &session);
+	if (rc)
+		return rc;
+
+	rc = serialargs_alloc_get_attributes(&ctrlargs, &template);
+	if (rc)
+		return rc;
+
+	if (serialargs_remaining_bytes(&ctrlargs)) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	/* Search objects only if no operation is on-going */
+	if (session_is_active(session)) {
+		rc = PKCS11_CKR_OPERATION_ACTIVE;
+		goto out;
+	}
+
+	if (session->find_ctx) {
+		EMSG("Active object search already in progress");
+		rc = PKCS11_CKR_FUNCTION_FAILED;
+		goto out;
+	}
+
+	rc = sanitize_client_object(&req_attrs, template,
+				    sizeof(*template) + template->attrs_size,
+				    PKCS11_UNDEFINED_ID, PKCS11_UNDEFINED_ID);
+	if (rc)
+		goto out;
+
+	/* Must zero init the structure */
+	find_ctx = TEE_Malloc(sizeof(*find_ctx), TEE_MALLOC_FILL_ZERO);
+	if (!find_ctx) {
+		rc = PKCS11_CKR_DEVICE_MEMORY;
+		goto out;
+	}
+
+	TEE_Free(template);
+	template = NULL;
+
+	switch (get_class(req_attrs)) {
+	case PKCS11_CKO_UNDEFINED_ID:
+	/* Unspecified class searches among data objects */
+	case PKCS11_CKO_SECRET_KEY:
+	case PKCS11_CKO_PUBLIC_KEY:
+	case PKCS11_CKO_PRIVATE_KEY:
+	case PKCS11_CKO_DATA:
+		break;
+	default:
+		EMSG("Find object of class %s (%"PRIu32") is not supported",
+		     id2str_class(get_class(req_attrs)),
+		     get_class(req_attrs));
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	/*
+	 * Scan all objects (sessions and persistent ones) and set a list of
+	 * candidates that match caller attributes. First scan all current
+	 * session objects (that are visible to the session). Then scan all
+	 * remaining persistent object for which no session object handle was
+	 * published to the client.
+	 */
+
+	LIST_FOREACH(obj, &session->object_list, link) {
+		if (check_access_attrs_against_token(session, obj->attributes))
+			continue;
+
+		if (req_attrs->attrs_count &&
+		    !attributes_match_reference(obj->attributes, req_attrs))
+			continue;
+
+		rc = find_ctx_add(find_ctx, pkcs11_object2handle(obj, session));
+		if (rc)
+			goto out;
+	}
+
+	LIST_FOREACH(obj, &session->token->object_list, link) {
+		uint32_t handle = 0;
+
+		if (check_access_attrs_against_token(session, obj->attributes))
+			continue;
+
+		/*
+		 * If there are no attributes specified, we return
+		 * every object
+		 */
+		if (req_attrs->attrs_count) {
+			rc = token_obj_matches_ref(req_attrs, obj);
+			if (rc == PKCS11_RV_NOT_FOUND)
+				continue;
+			if (rc != PKCS11_CKR_OK)
+				goto out;
+		}
+
+		/* Object may not yet be published in the session */
+		handle = pkcs11_object2handle(obj, session);
+		if (!handle) {
+			handle = handle_get(&session->object_handle_db, obj);
+			if (!handle) {
+				rc = PKCS11_CKR_DEVICE_MEMORY;
+				goto out;
+			}
+		}
+
+		rc = find_ctx_add(find_ctx, handle);
+		if (rc)
+			goto out;
+	}
+
+	if (rc == PKCS11_RV_NOT_FOUND)
+		rc = PKCS11_CKR_OK;
+
+	/* Save target attributes to search (if needed later) */
```

Comment:
maybe remove this inline comment. I don't find it really useful.

Reply:
Removed